### ISSUE TITLE
PRD: Multi-repository support for Enterprise users

### ISSUE THREAD

**Issue by toolmantim** [sourcegraph/cody#2628]
### Description:

As a user I want to add additional repositories that will be used as context in a Cody Chat in order to increase the quality of the chat responses (less hallucinations, higher chance of being able to answer, higher quality code recommendations)

### Objectives:

Enterprise user can chat with multiple repositories as context

### Acceptance Criteria:

- [ ] Can add and remove up to 10 additional repositories (fuzzy searching to find them)
- [ ] Show's inferred repository with tooltip
- [ ] Uses the additional repositories as context when Enhanced Context is enabled
- [ ]  When a user starts a new chat from an existing chat, the repository configuration is copied across to the new chat
- [ ] There is a "recently selected" group of repositories shown at the top of the quickpick (not shown in Figma), so you can easily keep re-selecting a common set of repositories in a new chat if the new chat was created not from an existing chat

Out of scope:

- Persisting the additional repositories between chats. Instead people will use the "recently selected" to choose the same ones as before if needed

### Design Tasks:

```[tasklist]
### Design Tasks
- [x] v1 design for comment from Eng and Prod - [Loom](https://www.loom.com/share/b96a8acfaabf47699e33c275297cecfd) - [Figma](https://www.figma.com/file/jEsnzgsf0hNuJbqB9pKaGm/VS-Code---Cody---Commands-up-front-and-ChatGPT++?type=design&node-id=1218-38734&mode=design&t=uzPYwXsMerPJlMSX-4)
```

### Engineering Tasks:

```[tasklist]
### Engineering Tasks
- [ ] https://github.com/sourcegraph/cody/issues/2622
- [ ] https://github.com/sourcegraph/cody/issues/2623
- [ ] https://github.com/sourcegraph/cody/issues/2624
- [x] (WIP) Pre-populate repo picker with currently selected repos
- [x] (WIP) Automatically populate the workspace repo(s)
- [x] (WIP) Display the selected repos in the enhanced context selector
- [x] (WIP) Save the set of selected repos in the chat transcript
- [x] (WIP) Search the set of selected repos
- [ ] (WIP) Add a way to `(refresh)` button to the repo picker which refreshes the repo list
- [x] (WIP) Change the default context from "embeddings" to "blended"
- [x] (WIP) Wire up "remove" button to removing repos from context source
- [ ] https://github.com/sourcegraph/cody/issues/2850
- [x] (WIP) Limit to enterprise users
- [x] (WIP) Test behavior signing into a different account with a chat open, etc.
- [ ] https://github.com/sourcegraph/jetbrains/issues/258
- [ ] (JetBrains) Implement createQuickPick for the multi-repo picker
- [x] (WIP) Integrate multi-repo remote search into inline edit context
- [x] (WIP) Add text to repo picker title that they can select N more repositories
- [ ] https://github.com/sourcegraph/cody/issues/2859
- [x] CodebaseStatusProvider._updateCodebase_NoFire bugs, see TODOs
- [ ] (WIP) Tests
- [x] (WIP) Display accurate labels in QuickPick and ECS—last part of name is primary
- [ ] (WIP) When creating a new chat in CTA, use the existing chat's selected repos
- [x] (WIP) When creating a new chat, use the workspace's repos by default
```

### Related Docs:

- https://github.com/sourcegraph/sourcegraph/pull/59392
- https://github.com/sourcegraph/cody/issues/1675
- https://github.com/sourcegraph/cody/issues/101
- VS Code supports fuzzy searching of repositories, if we load the entire list in I think. This could be fine? Otherwise we will have to implement fuzzy search ourselves. See [editor-context.ts](https://github.com/sourcegraph/cody/blob/74bee4df53e46de0e63c790d1719a2c8b37e56be/vscode/src/editor/utils/editor-context.ts) for prior art.
- [RFC877](https://docs.google.com/document/d/1t3-NrlgYd1tdgbabN3YoPpumuP0WpFeOTp_B1C14Rnw/edit)

**Comment by dominiccooney** [comment]
The workflow for adding repos: https://www.loom.com/share/b0c071fcd91d4fda9a9e869532611f15?sid=bedab6dc-2c0b-4309-9b1e-e1cea1a8c418

**Comment by dominiccooney** [comment]
End to end multi context search roughly works in WIP: https://www.loom.com/share/79bd672abcf54ec7abad7fa511c5bb31?sid=cecf3252-9e9e-43bd-b488-ddae29e103a6

**Comment by dominiccooney** [comment]
I'm seeking input on various product and design questions [in this doc.](https://docs.google.com/document/d/1IUhlVsoV90yTiFE4daQKexlU6T6O48Z4YVxQV3kM0ZA/edit#heading=h.h4qqug4obmw3) @toolmantim @taylorsperry

**Comment by dominiccooney** [comment]
We will need https://github.com/sourcegraph/jetbrains/issues/258 done before we can realize this feature for JetBrains.

**Comment by dominiccooney** [comment]
Testing caveat: You can only select repos without embeddings for now, or you will hit an error `searching vector DB: Qdrant is disabled.` See [internal Slack thread.](https://sourcegraph.slack.com/archives/C05FY86PHLM/p1705976154493259)

**Comment by dominiccooney** [comment]
Question for you @taylorsperry , do we need to support non-git repos? Everything VSCode Cody has done to date has used the git remote metadata to get a repository name to use with the Sourcegraph instance.

**Comment by dominiccooney** [comment]
Loom update: https://www.loom.com/share/d7ca236087fd44988c35a5a0e9e1ae0f

**Comment by dominiccooney** [comment]
I'm closing this, but please pushback if I got this wrong.

Is it correct to assume the scope of this issue is just VSCode, and we have an issue in sourcegraph/jetbrains for IntelliJ?

For VSCode, we are iffy on this one acceptance criteria:

> There is a "recently selected" group of repositories shown at the top of the quickpick (not shown in Figma), so you can easily keep re-selecting a common set of repositories in a new chat if the new chat was created not from an existing chat.

We have that, but it is *the repositories in the roots of your workspace,* not "recently used." Is this OK, @taylorsperry? Please send it back to me if there's more we need here.
